std/experimental/typecons.d

```diff
@@ -1024,3 +1024,34 @@ pure nothrow @safe unittest
     static assert(!__traits(compiles, arr ~= 4));
     assert((arr ~ 4) == [1, 2, 3, 4]);
 }
+
+// issue 17270
+pure nothrow @nogc @system unittest
+{
+    int i = 1;
+    Final!(int*) fp = &i;
+    assert(*fp == 1);
+    static assert(!__traits(compiles,
+        fp = &i // direct assignment
+    ));
+    static assert(is(typeof(*fp) == int));
+    *fp = 2; // indirect assignment
+    assert(*fp == 2);
+    int* p = fp;
+    assert(*p == 2);
+}
+
+pure nothrow @system unittest
+{
+    Final!(int[]) arr;
+    static assert(!__traits(compiles,
+        arr.length++
+    ));
+    arr.length = 10;
```

Comment:
This should not work; allowing modification of array length shouldn't be allowed. Or at least, if it's allowed, then modifying `.ptr` should also be allowed.  Since this is a bug, it should not be put into a unittest, because whatever is in a unittest implies that it's supposed to be allowed.

I would write the `arr.length` test as a not-compilable test, but commented out with a comment indicating the bug number.  But this is actually beyond the scope of this PR.  I would split out the array-related unittest into a separate PR, actually, so that we can merge the pointer fix now rather than later. Fixing array behaviour is something else that needs to be discussed, but shouldn't block the original intent of this PR.